### PR TITLE
Add validation data for PPO

### PR DESCRIPTION
**Pull request by gwarmstrong**
Allows specification of validation data with `--eval_data` flag for PPO. Expects a file in the same format as training data.

### CODE REVIEW

File: openrlhf/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -495,16 +519,27 @@ def save_logs_and_checkpoints(self, args, global_step, step_bar, logs_dict={}, c
...
+            logs_dict = self.evaluate(self.eval_dataloader, global_step)
+            if self._wandb is not None and self.strategy.is_rank_0():
+                logs = {
+                    "eval/%s" % k: v
+                    for k, v in {
+                        **logs_dict,
+                        "global_step": global_step,
+                    }.items()
+                }
+                # if self.experience_maker.perf_stats is not None:
```
Comment (Kipok) [inline openrlhf/trainer/ppo_trainer.py]: unused comments?

File: openrlhf/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -473,6 +476,27 @@ def training_step_critic(self, experience: Experience) -> Dict[str, float]:
...
+
+            pbar = tqdm(
+                range(dataloader.__len__()),
+                desc=f"Eval global step [{global_step}]",
+                disable=not self.strategy.is_rank_0(),
+            )
+
+            for prompts in dataloader:
+                for i, experience in enumerate(
+                    self.experience_maker.make_experience_list(prompts, **self.generate_kwargs)
```
Comment (Kipok) [inline openrlhf/trainer/ppo_trainer.py:490]: what happens if we use rloo that needs n_samples_per_prompt > 1? Will that mess up eval metric? If that's the case, perhaps we can hardcode reinforce in here

Comment (Kipok) [reply openrlhf/trainer/ppo_trainer.py:490]: Also are there any tracked statistics in the self.experience_maker? If so, eval might mess that up, but hopefully there are no such things

File: openrlhf/trainer/ray/ppo_actor.py [hunk 3]
```diff
@@ -349,6 +349,29 @@ def prepare_datasets(self):
...
+            eval_data = blending_datasets(
+                args.eval_data,
+                args.eval_data_probs,
+                strategy,
+                args.seed,
+                return_eval=False,
+            )
+
+            if args.limit_val_batches is not None:
+                samples_per_batch = args.rollout_batch_size // strategy.world_size
```
Comment (titu1994) [inline openrlhf/trainer/ray/ppo_actor.py:362]: Should this value be changed due to eval ? Wont it mess up training step too ?

**Review by titu1994 — COMMENTED** [review]
Looks good, but would want this on SFT too. Sft right now doesn't allow separate test data loader, it falls back to train data loader if set to -1

File: openrlhf/trainer/ppo_trainer.py (hunk 2 above)
Comment (gwarmstrong) [reply openrlhf/trainer/ppo_trainer.py:490]: > what happens if we use rloo that needs n_samples_per_prompt > 1? Will that mess up eval metric? If that's the case, perhaps we can hardcode reinforce in here

what do you mean by "mess up eval metric"? In this case, eval would take `n_samples_per_prompt` from the generate kwargs, and the eval reward would be the average of all  samples. Do you think it should be different?

> Also are there any tracked statistics in the self.experience_maker? If so, eval might mess that up, but hopefully there are no such things

There does appear to be an option for tracked perf stats, but [they are reset on every call to `make_experience_list`](https://github.com/OpenRLHF/OpenRLHF/blob/5e233c7a143a043b3263d5480a2dfa4fe5bfce8b/openrlhf/trainer/ppo_utils/experience_maker.py#L503-L508). There do not appear to be other meaningful stateful changes to the `experience_maker`, so I think this is safe.

File: openrlhf/trainer/ppo_trainer.py (hunk 1 above)
Comment (gwarmstrong) [reply openrlhf/trainer/ppo_trainer.py]: yeah I can remove unless you see a good reason we need perf stats for the eval experience maker

File: openrlhf/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -495,16 +519,27 @@ def save_logs_and_checkpoints(self, args, global_step, step_bar, logs_dict={}, c
...
+            if self._wandb is not None and self.strategy.is_rank_0():
+                logs = {
+                    "eval/%s" % k: v
+                    for k, v in {
+                        **logs_dict,
+                        "global_step": global_step,
+                    }.items()
+                }
+                # if self.experience_maker.perf_stats is not None:
+                #     logs.update({f"perf/experience_maker/{k}": v for k, v in self.experience_maker.perf_stats.items()})
```
Comment (gwarmstrong) [inline openrlhf/trainer/ppo_trainer.py]:
```suggestion
```

File: openrlhf/trainer/ray/ppo_actor.py (hunk 3 above)
Comment (gwarmstrong) [reply openrlhf/trainer/ray/ppo_actor.py:362]: Which value? `samples_per_batch`? It is only used here. And mess up how?

**Comment by gwarmstrong** [comment]
> Looks good, but would want this on SFT too.

@titu1994 makes sense, I'll add it in follow-up
